crates/rspack_plugin_mf/src/container/module_federation_runtime_plugin.rs

```diff
@@ -1,122 +1,78 @@
+  options: ModuleFederationRuntimePluginOptions,
 }
 
-#[async_trait]
-impl RuntimeModule for FederationRuntimeModule {
-  fn name(&self) -> Identifier {
-    self.id
-  }
-
-  fn attach(&mut self, chunk: ChunkUkey) {
-    self.chunk = Some(chunk);
-  }
-
-  fn stage(&self) -> RuntimeModuleStage {
-    RuntimeModuleStage::Normal
-  }
-
-  async fn generate(&self, compilation: &Compilation) -> rspack_error::Result<String> {
-    let chunk = compilation
-      .chunk_by_ukey
-      .expect_get(&self.chunk.expect("The chunk should be attached."));
-    Ok(federation_runtime_template(chunk, compilation).await)
+impl ModuleFederationRuntimePlugin {
+  pub fn new(options: ModuleFederationRuntimePluginOptions) -> Self {
+    Self::new_inner(options)
   }
 }
 
-async fn federation_runtime_template(chunk: &Chunk, compilation: &Compilation) -> String {
-  let federation_global = format!("{}.federation", RuntimeGlobals::REQUIRE);
-
-  let condition_map =
-    compilation
-      .chunk_graph
-      .get_chunk_condition_map(&chunk.ukey(), compilation, chunk_has_js);
-  let has_js_matcher = compile_boolean_matcher(&condition_map);
-
-  let chunk_matcher = if matches!(has_js_matcher, BooleanMatcher::Condition(false)) {
-    String::from("")
-  } else {
-    format!(
-      r#"
-chunkMatcher: function(chunkId) {{
-    return {has_js_matcher};
-}},
-"#,
-      has_js_matcher = &has_js_matcher.render("chunkId")
-    )
-  };
-
-  // Calculate rootOutputDir similar to webpack
-  let root_output_dir = {
-    let filename = get_js_chunk_filename_template(
-      chunk,
-      &compilation.options.output,
-      &compilation.chunk_group_by_ukey,
-    );
-    let output_name = compilation
-      .get_path(&filename, Default::default())
-      .await
-      .expect("failed to get output path");
-    get_undo_path(
-      &output_name,
-      compilation.options.output.path.to_string(),
-      false,
-    )
-  };
-
-  let root_output_dir_str = format!(
-    r#"rootOutputDir: "{}",
-"#,
-    root_output_dir
-  );
-
-  format!(
-    r#"
-if(!{federation_global}){{
-    {federation_global} = {{
-        {chunk_matcher}{root_output_dir_str}
-    }};
-}}
-"#,
-    federation_global = federation_global,
-    chunk_matcher = chunk_matcher,
-    root_output_dir_str = root_output_dir_str
-  )
-}
-
-#[plugin]
-#[derive(Debug, Default)]
-pub struct ModuleFederationRuntimePlugin;
-
 #[plugin_hook(CompilationAdditionalTreeRuntimeRequirements for ModuleFederationRuntimePlugin)]
 async fn additional_tree_runtime_requirements(
   &self,
   compilation: &mut Compilation,
   chunk_ukey: &ChunkUkey,
   _runtime_requirements: &mut RuntimeGlobals,
 ) -> Result<()> {
+  // Add base FederationRuntimeModule
   compilation.add_runtime_module(chunk_ukey, Box::<FederationRuntimeModule>::default())?;
+
+  Ok(())
+}
+
+#[plugin_hook(CompilerFinishMake for ModuleFederationRuntimePlugin)]
+async fn finish_make(&self, compilation: &mut Compilation) -> Result<()> {
```

Comment:
instead of injecting a entrypoint or patching entry dependencies into each entrypoint. We will use a dependency to addInclude the module. 

This will inject it with no require calls, so that our runtime module can handle startup of federation ahead of the entrypoint.